### ISSUE TITLE
salt-ssh fails with IPv6 address

### ISSUE THREAD

**Issue by jfindlay** [saltstack/salt#30561]
### ssh roster

``` yaml
jmoney-main ~ master # cat /etc/salt/roster
jmoney-rhel-7: {host: '2001:4800:7818:104:be76:4eff:fe04:618a', priv: /root/.ssh/id_RS,
  user: root}
```
### IPv6 address fails for normal `salt-ssh`

``` yaml
jmoney-main ~ master # salt-ssh jmoney-rhel-7 cmd.run 'ls -a'
jmoney-rhel-7:
    ----------
    retcode:
        11
    stderr:
        WARNING: Unable to locate current thin  version: /tmp/.root_bc136f_salt/version.
    stdout:
        ERROR: Failure deploying thin: /usr/bin/scp
        _edbc7885e4f9aac9b83b35999b68d015148caf467b78fa39c05f669c0ff89878
        deploy

        WARNING: Unable to locate current thin  version: /tmp/.root_bc136f_salt/version.
```
### but succeeds for `salt-ssh -r`

``` yaml
jmoney-main ~ master # salt-ssh jmoney-rhel-7 -r 'ls -a'
jmoney-rhel-7:
    ----------
    retcode:
        0
    stderr:
    stdout:
        .
        ..
        .bash_logout
        .bash_profile
        .bashrc
        .cshrc
        .ssh
        .tcshrc
```
### versions

``` yaml
jmoney-main ~ master # salt-ssh --versions
Salt Version:
           Salt: 2015.8.3-238-ge425cbd

Dependency Versions:
         Jinja2: 2.7.3
       M2Crypto: 0.21.1
           Mako: 1.0.0
         PyYAML: 3.11
          PyZMQ: 14.4.0
         Python: 2.7.9 (default, Mar  1 2015, 12:57:24)
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5
           cffi: 0.8.6
       cherrypy: Not Installed
       dateutil: Not Installed
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
        libgit2: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.2
   mysql-python: Not Installed
      pycparser: 2.10
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: Not Installed
          smmap: 0.8.2
        timelib: Not Installed

System Versions:
           dist: debian 8.2
        machine: x86_64
        release: 3.16.0-4-amd64
         system: debian 8.2
```

**Comment by The-Loeki** [comment]
Guess what happens when you execute

```
2016-02-14 11:42:44,974 [salt.client.ssh.shell][DEBUG][4621] Executing command: scp   -o KbdInteractiveAuthentication=no -o PasswordAuthentication=no -o GSSAPIAuthentication=no -o ConnectTimeout=90 -o Port=22 -o IdentityFile=/etc/salt/pki/master/ssh/salt-ssh.rsa -o User=root  /var/cache/salt/master/thin/thin.tgz 2001:4800:7818:104:be76:4eff:fe04:618a:/tmp/.root_03f252_salt/salt-thin.tgz
```

It's a pickle though, because you actually need to target `[2001:4800:7818:104:be76:4eff:fe04:618a]:/tmp/.root_03f252_salt/salt-thin.tgz` for this to work.

*BUT* that's a syntax SSH won't like (`ssh [2001:4800:7818:104:be76:4eff:fe04:618a]` won't work).

So methinks pretty much the only way around it is to detect an IPv6 address and add the brackets in `send` when necessary. Or use DNS like the rest of the decent world ;)

**Comment by Arabus** [comment]
Just stumbled across this also with the same issue. When running `salt-ssh -r` the thin version is not scped and everything is fine. 
The moment one wants the thin version though either the scp fails if the ipv6 address is specified in the roster without brackets or ssh fails because the address is specified with brackets.

#22984 seems to be the same issue.

I'd expect salt-ssh to detect ipv6 addresses and handle them properly or as a dirty hack at least allow setting ipv6: True and then handle the address accordingly

**Comment by Arabus** [comment]
The problem stems from https://github.com/saltstack/salt/blob/develop/salt/client/ssh/shell.py#L323 which does not detect IPv6 addresses and thus uses them verbatim from the roster file instead of adding the extra braces.

**Comment by Arabus** [comment]
Considering all the magic to simply run commands and copy stuff via ssh in the salt.client.ssh module I'd strongly suggest havong a look at existing implementations of ssh clients in python instead of reinventing the wheel

**Comment by pboos** [comment]
This would be nice to get fixed!

Current workaround is to just manually execute the scp command and add the [] around the ip. Afterwards everything works fine.

**Comment by The-Loeki** [comment]
Can be closed now that #38831 is in

**Comment by rallytime** [comment]
Thanks @The-Loeki!
